onnxruntime/core/optimizer/matmul_transpose_fusion.cc

```diff
@@ -0,0 +1,133 @@
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  std::deque<onnxruntime::NodeIndex> removed_nodes;
+
+  std::unordered_map<NodeArg*, size_t> consumer_count;
+
+  for (auto node_index : node_topology_list) {
+    auto& node = *graph.GetNode(node_index);
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    if ((!graph_utils::IsSupportedOptypeVersionAndDomain(node, "MatMul", {9}) &&
+         !graph_utils::IsSupportedOptypeVersionAndDomain(node, "TransposeMatMul", {9})) ||
+        !graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders())) {
+      continue;
+    }
+
+    NodeArg* left_input = node.MutableInputDefs()[0];
+    auto left = IsInputTranspose(graph, *left_input);
+
+    NodeArg* right_input = node.MutableInputDefs()[1];
+    auto right = IsInputTranspose(graph, *right_input);
+
+    if (!left.first && !right.first) {
+      continue;
+    }
+
+    if (left.first) {
+      size_t left_consumers = UpdateConsumerCount(graph, left_input, consumer_count);
+      if (left_consumers == 0)
+        removed_nodes.push_front(left.second->Index());
+      left_input = left.second->MutableInputDefs()[0];
+    }
+
+    if (right.first) {
+      size_t right_consumers = UpdateConsumerCount(graph, right_input, consumer_count);
+      if (right_consumers == 0)
+        removed_nodes.push_front(right.second->Index());
+      right_input = right.second->MutableInputDefs()[0];
+    }
+
+    const std::vector<NodeArg*> input_defs{left_input, right_input};
+    const std::vector<NodeArg*> output_defs{node.MutableOutputDefs()[0]};
+
+    Node& matmul_node = graph.AddNode(graph.GenerateNodeName("MatMul_With_Transpose"),
+                                      "TransposeMatMul",
+                                      "fused MatMul and Transpose ",
+                                      input_defs,
+                                      output_defs, {}, kMSDomain);
+    bool transpose_left = left.first;
+    if (node.OpType() == "TransposeMatMul") {
+      transpose_left ^= static_cast<bool>(node.GetAttributes().at("transA").i());
+    }
+    bool transpose_right = right.first;
+    if (node.OpType() == "TransposeMatMul") {
+      transpose_right ^= static_cast<bool>(node.GetAttributes().at("transB").i());
+    }
+    matmul_node.AddAttribute("transA", (int64_t)transpose_left);
+    matmul_node.AddAttribute("transB", (int64_t)transpose_right);
+    // Assign provider to this new node. Provider should be same as the provider for old node.
+    matmul_node.SetExecutionProviderType(node.GetExecutionProviderType());
+
+    graph_utils::FinalizeNodeFusion(graph, matmul_node, node);
+  }
+
+  // Have to remove node in reversed order for now to walk around the issue in RemoveNode
+  for (onnxruntime::NodeIndex removed_node : removed_nodes) {
+    graph.RemoveNode(removed_node);
+  }
+
```

Comment:
I wouldn't have expected this to be necessary. FinalizeNodeFusion can finalize replacing multiple nodes with one (and keep the edges in the graph valid whilst doing so). e.g. see onnxruntime\core\optimizer\conv_activation_fusion.cc

Reply:
tested with the FinalizeNodeFusion by following the recommended logic and bert toy training test is broken. will keep the code the same for now
Investigate in future as needed. 
@pengwa / @souptc - FYI

---
In reply to: [397608196](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397608196) [](ancestors = 397608196)